### DISCUSSION TITLE
In LINQ, the expression after key word "select" cannot get local varible before it

### DISCUSSION THREAD

**Discussion by martinrhan** [dotnet/csharplang#4937] [Language Ideas] [answered]
I have a piece of code like this.
```
        public static IEnumerable<T> FindInCircle<T>(this GameWorld gameWorld, Vector2 center, double range) where T : GameBody {
            return from GameBody body in gameWorld.world.CircleHitTest(center, range) where body is T select body as T;
        }
```
orginally, I wanted to write like this:
```
from GameBody body in gameWorld.world.CircleHitTest(center, range) where body is T t select t;
```
However it gave me Compiler Error CS0103, The name 't' does not exist in the current context.
Is it possible to get this improved?

**Comment by jnm2** [comment]
See: https://github.com/dotnet/csharplang/discussions/159, https://github.com/dotnet/csharplang/discussions/355

**Comment by HaloFour** [comment]
Unfortunately the team [rejected](https://github.com/dotnet/csharplang/blob/main/meetings/2018/LDM-2018-01-03.md#scoping-of-expression-variables-in-query-clauses) and effectively closed the door to allowing LINQ queries to behave as you suggest.  The pattern variables declared within a LINQ clause are not promoted to range variables and not available through the rest of the query.

There was a proposal at one point to allow pattern variables declared within the `let` clause to be promoted to a range variable, so that you could write:

```cs
from item in collection
let isT = item is T value
where isT
select value
```

But it seems that was never implemented and that probably couldn't be implemented without source breaking changes (albeit very minor ones).

Right now the best you can do is:

```cs
from item in collection
let match = item is T value ? (true, value) : (false, default)
where match.Item1
select match.Item2;
```

**Comment by Richiban** [comment, marked as answer]
In the spirit of trying to get you through your current use cases... don't forget about the `.OfType<>()` Linq method.

In the example you gave you'd use it like this:

```cs
public static IEnumerable<T> FindInCircle<T>(this GameWorld gameWorld, Vector2 center, double range) 
    where T : GameBody 
{
    return gameWorld.world.CircleHitTest(center, range).OfType<T>();
}
```
